src/Components/Components/src/RenderTree/EventHandlerDelegateWithClosuresComparer.cs

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Collections.Concurrent;
+using System.Reflection;
+using System.Linq.Expressions;
+
+namespace Microsoft.AspNetCore.Components.RenderTree
+{
+    internal static class EventHandlerDelegateWithClosuresComparer
+    {
+        private static readonly ConcurrentDictionary<Type, Func<object, object, bool>> ClosureComparerCache = new();
+
+        internal static bool EventCallBackEquals(ref EventCallback left, ref EventCallback right)
+        {
+            if (Equals(left, right))
+            {
+                return true;
+            }
+
+            // Whenever the receivers are different and they are explicit we assume that
+            // the EventCallbacks are not equal.
+            if (left.Receiver != right.Receiver &&
+                (left.RequiresExplicitReceiver || right.RequiresExplicitReceiver))
+            {
+                return false;
+            }
+
+            return DelegateEquals(left.Delegate, right.Delegate);
+        }
+
+        internal static bool DelegateEquals(MulticastDelegate? left, MulticastDelegate? right)
+        {
+            if (Equals(left, right))
+            {
+                return true;
+            }
+
+            // If any of the delegates or their targets are null (static delegates) then normal
+            // equality is sufficient. No additional testing required.
+            if (left == null || right == null || left.Target == null || right.Target == null)
+            {
+                return false;
+            }
+
+            var oldTargetType = left.Target!.GetType();
+            var newTargetType = right.Target!.GetType();
+
+            // if the types are not the same, or the methods are not the same, these possible closures are
+            // not the same anyway.
+            if (oldTargetType != newTargetType || left.Method != right.Method)
```

Comment:
Is there any possible case where `left.Method == right.Method` but `oldTargetType != newTargetType`? I'm unsure both checks are needed.

Reply:
You are right. The Method came from a unit test that actually failed, which made me realise that the Method should be checked as well. But you are right. If left.Method != right.Method, you are sure that the types will be different anyway. I'll change this. 

This BTW answers your other question: Yes, the compiler will put all lambda's inside the scope on one single instance sometimes when the captured variables are the same. I'm not sure it might even combine all free variables to a single closure type as optimization. I triggered this in a test though.